modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -854,51 +855,82 @@ def join(
             right_on
         ), "'left_on' and 'right_on' lengths don't match"
 
-        for col in left_on:
-            assert col in self.columns, f"'left_on' references unknown column {col}"
-        for col in right_on:
-            assert col in other.columns, f"'right_on' references unknown column {col}"
+        def validate(what, df, col_names):
+            cols = df.columns
+            idx_cols = []
+            for i, col in enumerate(col_names):
+                if col not in cols:
```

Comment:
We seem to have zero tests for merging on an index level in `test_join_sort.py`. I have a separate PR (#5594) where I'm trying to implement a general logic for precomputing resulting columns for `.merge()` operation. When joining on an index level, pandas may behave really tricky and sometimes even unpredictable in terms of how it forms the resulting columns. 

In my PR I'm also adding a test for merging on index. I've tried to run it on your branch and it's failing. It seems that it's easier to default to pandas in such cases for now than trying to partially support them.

<details><summary>test_merge_on_index</summary>

```python
def test_merge_on_index():
    pandas_df1 = pandas.DataFrame(
        {
            "idx_key1": [1, 2, 3, 4],
            "idx_key2": [2, 3, 4, 5],
            "idx_key3": [3, 4, 5, 6],
            "data_col1": [10, 2, 3, 4],
            "col_key1": [3, 4, 5, 6],
            "col_key2": [3, 4, 5, 6],
        }
    )

    pandas_df1 = pandas_df1.set_index(["idx_key1", "idx_key2"])
    modin_df1 = pd.DataFrame(pandas_df1)

    pandas_df2 = pandas.DataFrame(
        {
            "idx_key1": [4, 3, 2, 1],
            "idx_key2": [5, 4, 3, 2],
            "idx_key3": [6, 5, 4, 3],
            "data_col2": [10, 2, 3, 4],
            "col_key1": [6, 5, 4, 3],
            "col_key2": [6, 5, 4, 3],
        }
    )

    pandas_df2 = pandas_df2.set_index(["idx_key2", "idx_key3"])
    modin_df2 = pd.DataFrame(pandas_df2)

    for on in (
        ["col_key1", "idx_key1"],
        ["col_key1", "idx_key2"],
        ["col_key1", "idx_key3"],
        ["idx_key1"],
        ["idx_key2"],
        ["idx_key3"],
    ):
        res = modin_df1.merge(modin_df2, on=on)
        ref = pandas_df1.merge(pandas_df2, on=on)
        df_equals(res, ref)

    for left_on, right_on in (
        (["idx_key1"], ["col_key1"]),
        (["col_key1"], ["idx_key3"]),
        (["idx_key1"], ["idx_key3"]),
        (["idx_key2"], ["idx_key2"]),
        (["col_key1", "idx_key2"], ["col_key2", "idx_key2"]),
    ):
        res = modin_df1.merge(modin_df2, left_on=left_on, right_on=right_on)
        ref = pandas_df1.merge(pandas_df2, left_on=left_on, right_on=right_on)
        df_equals(res, ref)
```

</details>

Reply:
I think that your PR should be merged before #5594. Given all that, could you please add the test that I've mentioned into your PR and probably get back the logic that was defaulting to pandas on index join?

Reply:
Done.
The test passes in my env, but I can't understand what's happening with CI. After the last commit, CI failed with this error - https://github.com/modin-project/modin/actions/runs/4178452249/jobs/7237265084:
```
modin/pandas/test/dataframe/test_join_sort.py:758:1: F811 redefinition of unused 'test_merge_on_index' from line 333
```
The error is weird, because there is no line 758 in this file - https://github.com/AndreyPavlenko/modin/blob/issue-5576/modin/pandas/test/dataframe/test_join_sort.py. I renamed the test, force pushed, renamed it back and force pushed again - in both cases CI passed, as you can see below, but only with a single test case:
```
All checks have passed
1 successful check

docs/readthedocs.org:modin — Read the Docs build succeeded!
```

Reply:
@dchigarev Do you have an idea why flake8 fails? I can't reproduce the failure in my env:
```
flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py
modin/pandas/test/dataframe/test_join_sort.py:759:1: F811 redefinition of unused 'test_merge_on_index' from line 333
```

Reply:
maybe the master branch somehow incorrectly merges into your branch during CI? Can you try to rebase your branch on the latest master and force push it?

Reply:
I tried multiple times - it doesn't help.

Reply:
I can reproduce the flake8 failure by merging your branch to the latest master branch locally. Note that I'm merging your branch IN master, not backward.

Reply:
so, to be more precise the steps are:
1. `git fetch origin master`
2. `git checkout origin/master | git checkout -b test_flake8`
3. `git merge issue-5576`
4. `flake8 modin` - fails

Reply:
Thanks, I could reproduce it now.